### DISCUSSION TITLE
Add Trajectory Consistency Distillation (TCD)

### DISCUSSION THREAD

**Discussion by Nerual-Dreming** [lllyasviel/Fooocus#2733] [Ideas]
Hello. Please add a new generation mode and a sampler for it. It shows better results than lсm. 

https://huggingface.co/h1t/TCD-SDXL-LoRA

https://huggingface.co/docs/diffusers/main/en/using-diffusers/inference_with_tcd_lora

https://huggingface.co/docs/diffusers/main/en/api/schedulers/tcd#diffusers.TCDScheduler

**Comment by mashb1t** [comment]
also see https://github.com/lllyasviel/Fooocus/discussions/2416#discussioncomment-8674200

**Comment by mashb1t** [comment]
Implemented in https://github.com/lllyasviel/Fooocus/pull/2907, please check if the virtual scheduler makes sense or if we should rather use sgm_uniform.

**Reply by Nerual-Dreming** [reply]
> Implemented in #2907, please check if the virtual scheduler makes sense or if we should rather use sgm_uniform.

It seems that sigma actually works better. I also found that cfg 1.5 looks nicer than 1. And the optimal number of steps in terms of speed/quality ratio is 6 for TCD mode.

![image](https://github.com/lllyasviel/Fooocus/assets/135246066/3d3b77a3-e10e-4f66-a4a3-20efe0fdf35e)

**Reply by mashb1t** [reply]
@Nerual-Dreming are you referring to using TCD in combination with Hyper-SD or which performance do you use for cfg 1.5?

**Reply by Nerual-Dreming** [reply]
> @Nerual-Dreming are you referring to using TCD in combination with Hyper-SD or which performance do you use for cfg 1.5?

I used Performance Speed, but changed the following settings: 
CFG Mimicking from TSNR = 1
Sampler = tcd
Scheduler = sgm_uniform
Forced Overwrite of Sampling Step = 6
Guidance Scale = 1.5
I think it's an interesting result in this mode. I have attached some more examples below.
![2024-05-29_03-04-59_8425](https://github.com/lllyasviel/Fooocus/assets/135246066/66a274b2-4474-49d1-aeaa-2f5662b295f5)
![2024-05-29_03-06-28_9623](https://github.com/lllyasviel/Fooocus/assets/135246066/0b619fc7-fa48-4f41-9db9-3ae774bc9da9)
![2024-05-29_03-08-04_6262](https://github.com/lllyasviel/Fooocus/assets/135246066/fbc56993-66f0-4432-a7c3-4939ba197494)
![2024-05-29_03-10-50_6693](https://github.com/lllyasviel/Fooocus/assets/135246066/6a60a10a-73a3-475f-a2a9-20c6769416d8)
![171693827416587073](https://github.com/lllyasviel/Fooocus/assets/135246066/38591483-723a-4557-8c96-d450166e05a8)

![2024-05-29_03-00-31_2767](https://github.com/lllyasviel/Fooocus/assets/135246066/069a745b-9238-40bc-b831-39b547c2e219)
![2024-05-29_03-03-46_3488](https://github.com/lllyasviel/Fooocus/assets/135246066/cca1ec8e-54b0-4999-84b9-37fda6f40e70)

Maybe you know how to improve the image?

**Reply by mashb1t** [reply]
Going to check this out in the evening. Fooocus does not yet have a preset for Hyper-SD & TCD, so thank you for the insights.
Referenced in issue https://github.com/lllyasviel/Fooocus/issues/3044

**Reply by mashb1t** [reply]
@Nerual-Dreming would it be possible to provide the prompts used for your example images? Sadly no metadata is contained.
I strongly assume that you also use the TCD LoRA with weight 1, right?

**Reply by Nerual-Dreming** [reply]
> @Nerual-Dreming would it be possible to provide the prompts used for your example images? Sadly no metadata is contained. I strongly assume that you also use the TCD LoRA with weight 1, right?

Yep, I attach info text.

```
{
"prompt": "intricate shading, three-dimensional appearance, central figure fox surrounded by various flowers, soft pastel background colors, assortment of flowers complementing the fox\u2019s orange fur, meticulously detailed fur, soft and hard shading, plush glossy texture, semi-realistic style, vibrant colors, intricate detail, charming subject matter, appealing to wildlife art enthusiasts, digital illustration techniques, inspired by artists like Sakimichan or TsaoShin",
"negative_prompt": "",
"prompt_expansion": "intricate shading, three-dimensional appearance, central figure fox surrounded by various flowers, soft pastel background colors, assortment of flowers complementing the fox\u2019s orange fur, meticulously detailed fur, soft and hard shading, plush glossy texture, semi-realistic style, vibrant colors, intricate detail, charming subject matter, appealing to wildlife art enthusiasts, digital illustration techniques, inspired by artists like Sakimichan or TsaoShin, magic celestial, very coherent, cute quality, extremely aesthetic, deep complimentary color, lovely atmosphere, enhanced light, novel composition, beautiful dramatic cinematic perfect surreal dynamic highly complex scenic full flowing, massive elaborate epic breathtaking, stunning, spectacular fine artistic balance, creative, positive spiritual, symbolic, radiant, magical, pure",
"styles": "['Fooocus V2', 'Fooocus Masterpiece']",
"performance": "Speed",
"steps": 6,
"resolution": "(1280, 768)",
"guidance_scale": 1.58,
"sharpness": 0,
"adm_guidance": "(1, 0.8, 0.3)",
"base_model": "juggernautXL_v9Rundiffusionphoto2.safetensors",
"refiner_model": "None",
"refiner_switch": 0.5,
"adaptive_cfg": 1,
"clip_skip": 2,
"sampler": "tcd",
"scheduler": "sgm_uniform",
"vae": "Default (model)",
"seed": "1092663982407441428",
"lora_combined_1": "FAST\\TCD-SpeedUp-SDXL-PAseer.safetensors : 1.0",
"metadata_scheme": false,
"version": "Fooocus v2.4.1"
}
```
```
{
"prompt": "A breathtaking image of a white human skull with liquid gold being poured on its head, highly detailed, high quality surreal image, movie poster style",
"negative_prompt": "",
"prompt_expansion": "A breathtaking image of a white human skull with liquid gold being poured on its head, highly detailed, high quality surreal image, movie poster style, dynamic composition, dramatic bright colors, cinematic atmosphere, inspired, rich deep color, intricate, beautiful, epic, stunning, creative, perfect scientific, sharp focus, futuristic, thought complex, amazing fine detail, professional background, pretty cute",
"styles": "['Fooocus V2', 'Fooocus Masterpiece']",
"performance": "Speed",
"steps": 6,
"resolution": "(1280, 768)",
"guidance_scale": 1.58,
"sharpness": 0,
"adm_guidance": "(1, 0.8, 0.3)",
"base_model": "juggernautXL_v9Rundiffusionphoto2.safetensors",
"refiner_model": "None",
"refiner_switch": 0.5,
"adaptive_cfg": 1,
"clip_skip": 2,
"sampler": "tcd",
"scheduler": "sgm_uniform",
"vae": "Default (model)",
"seed": "1067485533856333156",
"lora_combined_1": "FAST\\TCD-SpeedUp-SDXL-PAseer.safetensors : 1.0",
"metadata_scheme": false,
"version": "Fooocus v2.4.1"
}
```

```
{
"prompt": "soft pastel colors, cartoon style illustration of a woman as she sees the world while experiencing hallucinations, stoned, splash art, splashed pastel colors, (soft iridiscent glowy smoke) motion effects, best quality, wallpaper art, UHD, centered image, MSchiffer art, ((flat colors)), (cel-shading style) very vibrant neon colors, ((low saturation)) ink lines, iridiscent",
"negative_prompt": "",
"prompt_expansion": "soft pastel colors, cartoon style illustration of a woman as she sees the world while experiencing hallucinations, stoned, splash art, splashed pastel colors, (soft iridiscent glowy smoke) motion effects, best quality, wallpaper art, UHD, centered image, MSchiffer art, ((flat colors)), (cel-shading style) very vibrant neon colors, ((low saturation)) ink lines, iridiscent, highly detailed, intricate, innocent, fine detail, dramatic light, complete whole background, cinematic, stunning, romantic, beautiful dynamic, glowing, rich deep color, perfect complex amazing composition, sharp focus, elegant, epic, full bright, gorgeous, appealing, creative, atmosphere, ideal, great luxury, inspiring",
"styles": "['Fooocus V2', 'Fooocus Masterpiece']",
"performance": "Speed",
"steps": 6,
"resolution": "(1280, 768)",
"guidance_scale": 2,
"sharpness": 0,
"adm_guidance": "(1, 0.8, 0.3)",
"base_model": "juggernautXL_v9Rundiffusionphoto2.safetensors",
"refiner_model": "None",
"refiner_switch": 0.5,
"adaptive_cfg": 1,
"clip_skip": 2,
"sampler": "tcd",
"scheduler": "sgm_uniform",
"vae": "Default (model)",
"seed": "5179042452506990297",
"lora_combined_1": "FAST\\TCD-SpeedUp-SDXL-PAseer.safetensors : 1.0",
"lora_combined_2": "Best_LoRas_Mar24\\xl_more_art-full_v1.safetensors : 0.3",
"metadata_scheme": false,
"version": "Fooocus v2.4.1"
}
```
